### ISSUE TITLE
Update BCryptSetProperty to optionally take pbinput

### ISSUE THREAD

**Issue by tofay** [microsoft/windows-rs#3364]
### Suggestion

Regarding https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Security/Cryptography/fn.BCryptSetProperty.html, for a HKDF operation I needed to set a property to a null pointer (I tried setting an empty array but it caused a CNG error).

I ended up just not using windows-rs for the one call in my crate that needed to set a null pointer: https://github.com/tofay/rustls-cng-crypto/blob/2f1de485a6856a8db4ccadb073e5ec3cf872d04f/src/hkdf.rs#L97.

**Comment by kennykerr** [comment]
From the header:

```C++
//  - Finally the key (or secret) is finalized via the
//  BCRYPT_HKDF_PRK_AND_FINALIZE property. In this case the input property
//  must be NULL since the PRK was already passed in.
```

Yes, I think we should be able to make this work - I'll see what I can do.

**Comment by tim-weis** [comment]
That's a bug, albeit not in the `windows` crate.

The *bcrypt.h* SDK header file declares `BCryptSetProperty` like this:

```c
_Must_inspect_result_
NTSTATUS
WINAPI
BCryptSetProperty(
    _Inout_                 BCRYPT_HANDLE   hObject,
    _In_z_                  LPCWSTR pszProperty,
    _In_reads_bytes_(cbInput)    PUCHAR   pbInput,
    _In_                    ULONG   cbInput,
    _In_                    ULONG   dwFlags);
```

The [win32metadata](https://github.com/microsoft/win32metadata) tooling faithfully transforms this into

```csharp
public static extern NTSTATUS BCryptSetProperty(
  [In, Out] BCRYPT_HANDLE hObject,
  [Const, In] PWSTR pszProperty,
  [MemorySize(BytesParamIndex = 3), In] byte* pbInput,
  [In] uint cbInput,
  [In] uint dwFlags);
```

This gets picked up by the `windows` crate's code generator that collapses `pbInput` and `cbInput` into a slice. Rust [mandates](https://doc.rust-lang.org/core/slice/fn.from_raw_parts.html) that even zero-length slices are constructed from (and represented with) a valid (i.e., non-null) pointer. All of this is very much in line with the API contract spelled out via the SAL annotations.

The interesting bit here is the `_In_reads_bytes_(cbInput)` SAL annotation that requires `pbInput` to be a valid pointer to an array of size `cbInput` (as opposed to [`_In_reads_bytes_opt_`](https://learn.microsoft.com/en-us/cpp/code-quality/annotating-function-parameters-and-return-values#optional-pointer-parameters) that would accept a null pointer as well).

The bug could be either one of:

* The `BCryptSetProperty` implementation that exhibits different behavior for a zero-length input array depending on whether it is expressed through a null pointer or a non-null pointer.
* The SAL annotation for the function declaration that mandates a non-null pointer always (mind you, passing a null pointer won't pass static code analysis when called from C either).
* Client code using the API out-of-spec (the comment quoted by Kenny earlier suggests that this isn't the case).

*Properly* fixing this would necessitate a change to the header file (and possibly the implementation) of the BCrypt library. That'll take a hot minute to trickle down. What you can do in the meantime is taking advantage of the [`windows-targets` crate](https://kennykerr.ca/rust-getting-started/understanding-windows-targets.html)'s [`link!`](https://docs.rs/windows-targets/0.52.6/windows_targets/macro.link.html) macro to generate the raw import binding, as the `windows` create is doing:

```rust
link!("bcrypt.dll" "system" fn BCryptSetProperty(hobject : BCRYPT_HANDLE, pszproperty : windows_core::PCWSTR, pbinput : *const u8, cbinput : u32, dwflags : u32) -> super::super::Foundation:: NTSTATUS);
```

**Comment by riverar** [comment]
Agree, looks like a bcrypt bug; makes you wonder how they satisfied the SAL tests.

We can take this to win32metadata and get the `input` marked as optional to support nullable properties.

**Comment by tim-weis** [comment]
> \[...\] makes you wonder how they satisfied the SAL tests.

Checking an `_In_` argument against a null pointer is valid, even if superfluous. **Not checking** an `_In_opt_` pointer against null would fail, but `BCryptSetProperty` isn't using the `_opt_`-variant. The expectation here is that SAL testing will succeed, and that is valid (in face of an incorrect/inappropriate SAL annotation).
